### PR TITLE
CFY-4711 added subtree support and .travis yaml

### PR DESCRIPTION
**Pull request by adamlavie**

### CODE REVIEW

File: authentication/ldap_authentication_provider.py [hunk 1]
```diff
@@ -12,35 +12,147 @@
...
-        username, password = utils.get_basic_http_authentication_info()
+        who, password = self.get_creds_from_request()
+        if self.is_subtree_authentication_method():
+            self.authenticate_by_attribute(who,
+                                           password,
+                                           default_attr=DEFAULT_AUTH_ATTR_NAME)
+        elif self.is_dn_authentication_method():
+            self.try_bind_s(who, password)
+        else:
+            raise RuntimeError('Failed determining authentication method. '
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: this code is probably not reachable, cause if the no user_id_attribute was given we refer to the 'who' as a DN, and if other values are missing (like the admin details) we should raise a value error on that earlier

File: authentication/active_directory_authentication_provider.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    # Active Directory authentication can be done by subtree match or according
+    # to the UPN attribute which is the default auth attr in Active Directory.
+    def authenticate(self, userstore=None):
+        who, password = self.get_creds_from_request()
+
+        if self.is_subtree_authentication_method():
+            self.authenticate_by_attribute(who,
+                                           password,
+                                           default_attr=DEFAULT_AUTH_ATTR_NAME)
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: the default should be set as the default value in the method signature instead of passed here. Also, pass in self.user_id_attribute instead of DEFAULT_AUTH_ATTR_NAME

File: authentication/ldap_authentication_provider.py [hunk 3]
```diff
@@ -12,35 +12,147 @@
...
+                               'LDAP configuration is invalid.')
+        return who
+
+    def authenticate_by_attribute(self, who, password, default_attr=None):
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: 1. should probably be renamed to "authenticate_by_binding", attribute doesn't mean much in that context
2. default_attr should be renamed "identifying_attr" and the default value should be DEFAULT_AUTH_ATTR_NAME instead of None

File: authentication/ldap_authentication_provider.py [hunk 4]
```diff
@@ -12,35 +12,147 @@
...
+            conn = self.synchronous_bind(self.admin_user_id,
+                                         self.admin_password)
+        except AuthenticationException as e:
+            err = 'Failed binding to \'{0}\' using the dedicated ' \
+                  'admin_dn: {1}, provided in the configuration; {2}'\
+                .format(self.ldap_url, who, e)
+            self.lgr.warning(err)
+            raise RuntimeError(err)
+
+        attribute_name = self.user_id_attribute or default_attr
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: self.user_id_attribute should be passed as the identifying_attr arg to this method instead of this line

File: authentication/ldap_authentication_provider.py [hunk 5]
```diff
@@ -12,35 +12,147 @@
...
+                .format(self.base_dn, search_filter, e)
+            self.handle_authentication_error(err, conn)
+
+        try:
+            # Retrieve query results according to the result ID in order to
+            # obtain the user DN to bind with.
+            while True:
+                result_type, result_data = conn.result(ldap_result_id, 0)
+                if result_data:
+                    if result_type == ldap.RES_SEARCH_ENTRY:
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: and if result_type is different? what does it mean? how should we respond?

File: authentication/ldap_authentication_provider.py [hunk 6]
```diff
@@ -12,35 +12,147 @@
...
+            err = 'Failed searching for user under base DN: {0}, according ' \
+                  'to the following attribute filter pattern {1}; {2}'\
+                .format(self.base_dn, search_filter, e)
+            self.handle_authentication_error(err, conn)
+
+        try:
+            # Retrieve query results according to the result ID in order to
+            # obtain the user DN to bind with.
+            while True:
+                result_type, result_data = conn.result(ldap_result_id, 0)
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: 1. does that retrieve one record at a time? does that mean you are able to read more than one record? the passed value should have been unique (like user ID) so multiple values should be considered invalid here. I see now that you check it later but there is not point in appending who-knows-how-many-records, when we can already raise an exception here when we find there is more than one.
2. I think there is a better API for that named "find" instead of "search", not sure about it but please look.

File: authentication/ldap_authentication_provider.py [hunk 7]
```diff
@@ -12,35 +12,147 @@
...
+                    break
+        except ldap.LDAPError as e:
+            err = 'Failed retrieving results according to result_id; {0}'\
+                .format(e)
+            self.handle_authentication_error(err, conn)
+
+        # unbind admin connection.
+        conn.unbind()
+
+        if len(result_list) != 1:
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: see comment above

File: authentication/ldap_authentication_provider.py [hunk 8]
```diff
@@ -12,35 +12,147 @@
...
+            err = 'Expecting a single result matching the user with username' \
+                  ' {0}. Found {1}.'.format(who, len(result_list))
+            self.handle_authentication_error(err, conn)
+
+        # extract the user DN from the search result.
+        user_dn = result_list.pop()[0][0]
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: a bit strange, no?

File: authentication/ldap_authentication_provider.py [hunk 9]
```diff
@@ -12,35 +12,147 @@
...
+
+        # attempt to bind using the matching user DN and provided password.
+        self.try_bind_s(user_dn, password)
+
+    def handle_authentication_error(self, err, conn):
+        conn.unbind()
+        self.lgr.warning(err)
+        raise AuthenticationException(err)
+
+    def try_bind_s(self, who, password):
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: maybe rename to validate_bind, since "try_bind_s" does not explain what it really does (binds just to verify the credentials and then unbinds)

File: authentication/ldap_authentication_provider.py [hunk 10]
```diff
@@ -12,35 +12,147 @@
...
-            conn.unbind_s()
+            conn.bind_s(who, password)
+        except ldap.INVALID_CREDENTIALS as e:
+            err = 'Invalid credentials specified for user: \'{0}\''\
+                .format(who)
+            self.handle_authentication_error(err, conn)
+        except ldap.LDAPError as e:
+            err = 'Failed to bind user \'{0}\'; {1}' \
+                .format(who, e)
+            self.handle_authentication_error(err, conn)
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: finally:
  conn.unbind()

Comment (noakup) [inline authentication/ldap_authentication_provider.py]: finally:
  conn.unbind()

Comment (noakup) [inline authentication/ldap_authentication_provider.py]: since you don't really need the connection, why not:
1. rename the function to validate_connection
2. add a finally here to unbind

Comment (noakup) [inline authentication/ldap_authentication_provider.py]: since you don't really need the connection, why not:
1. rename the function to validate_connection
2. add a finally here to unbind

File: authentication/ldap_authentication_provider.py [hunk 11]
```diff
@@ -12,35 +12,147 @@
...
+
+        return conn
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: no need to return it

File: authentication/ldap_authentication_provider.py [hunk 12]
```diff
@@ -12,35 +12,147 @@
...
+            self.handle_authentication_error(err, conn)
+
+        # extract the user DN from the search result.
+        user_dn = result_list.pop()[0][0]
+
+        # attempt to bind using the matching user DN and provided password.
+        self.try_bind_s(user_dn, password)
+
+    def handle_authentication_error(self, err, conn):
+        conn.unbind()
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: the connection might not have been establish, should probably wrap the unbind with a try-except clause

File: authentication/ldap_authentication_provider.py [hunk 13]
```diff
@@ -12,35 +12,147 @@
...
+        self.lgr.warning(err)
+        raise AuthenticationException(err)
+
+    def try_bind_s(self, who, password):
+            conn = self.synchronous_bind(who, password)
+            conn.unbind()
+
+    def synchronous_bind(self, who, password):
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: since synchronous is the only way we work, maybe no need to name it here. it's a bit confusing in my opinion (yes I am aware of the async API but we're not using it)

File: authentication/ldap_authentication_provider.py [hunk 14]
```diff
@@ -12,35 +12,147 @@
...
+        self.admin_user_id = admin_user_id
+        self.admin_password = admin_password
+        self.user_id_attribute = user_id_attribute
 
+    # LDAP authentication can be done by subtree match or according to
+    # dn attribute.
     def authenticate(self, userstore=None):
-        username, password = utils.get_basic_http_authentication_info()
+        who, password = self.get_creds_from_request()
+        if self.is_subtree_authentication_method():
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: 1. the authentication method is not sub-tree, it's actually user-binding in both cases, but the difference as I see it is that the DN is not given, and so we must search for the user object first and retrieve the DN from it. I suggest we cancel the  2 seemingly different auth methods (since in fact authentication is always by binding the user using its DN). instead, we should check if the given user_id_attribute is other the DN (or none) and if so - retrieve the user DN by it, through a method named 'get_user_by_attribute' and 'get_DN_from_user_object'.
2. it should work even if no sub-tree was supplied, I know that's not the example Uri gave you but generally if no sub-tree is given the root is expected to be used, that's the default...

File: authentication/ldap_authentication_provider.py [hunk 15]
```diff
@@ -12,35 +12,147 @@
...
+            self.authenticate_by_attribute(who,
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: should be renamed as well, see comment on message signature

File: authentication/active_directory_authentication_provider.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    def __init__(self, domain_name=None, **kwargs):
+        super(ActiveDirectoryAuthenticationProvider, self).__init__(**kwargs)
+        self.domain_name = domain_name
+
+    # Active Directory authentication can be done by subtree match or according
+    # to the UPN attribute which is the default auth attr in Active Directory.
+    def authenticate(self, userstore=None):
+        who, password = self.get_creds_from_request()
+
+        if self.is_subtree_authentication_method():
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: rename (see below)

File: authentication/active_directory_authentication_provider.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+            self.authenticate_by_attribute(who,
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: rename (see below)

File: authentication/active_directory_authentication_provider.py [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+    # By default, active directory allows for authentication using the UPN
+    # attribute. If no base_nd specified, UPN authentication will be used.
+    def authenticate_by_upn(self, who, password):
+        if self.domain_name:
+            upn = '{0}@{1}'.format(who, self.domain_name)
+        else:
+            upn = who
+
+        self.try_bind_s(upn, password)
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: rename to 'validate_bind' (see below)

File: authentication/active_directory_authentication_provider.py [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+    # to the UPN attribute which is the default auth attr in Active Directory.
+    def authenticate(self, userstore=None):
+        who, password = self.get_creds_from_request()
+
+        if self.is_subtree_authentication_method():
+            self.authenticate_by_attribute(who,
+                                           password,
+                                           default_attr=DEFAULT_AUTH_ATTR_NAME)
+        elif self.is_authenticate_by_upn_method():
+            self.lgr.debug('Attempting to connect using the default active '
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: The message should be extra clear, so use the full and clear name here, this is probably User principal name or something like it

File: authentication/ldap_authentication_provider.py [hunk 20]
```diff
@@ -12,35 +12,147 @@
...
-            return username
-        except Exception as e:
-            raise Exception(
-                'Failed to authenticate user {0}; {1}'
-                .format(username, str(e)))
-        finally:
-            conn.unbind_s()
+            conn.bind_s(who, password)
+        except ldap.INVALID_CREDENTIALS as e:
+            err = 'Invalid credentials specified for user: \'{0}\''\
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: credential are given per request, so 'specified' ->> 'given' or 'supplied'

File: authentication/active_directory_authentication_provider.py [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+                           'directory UPN {0}@{1}'.format(self.admin_user_id,
+                                                          self.domain_name))
+            self.authenticate_by_upn(who, password)
+        else:
+            raise RuntimeError('Failed determining authentication method. '
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: this code is probably not reachable, cause if the no user_id_attribute was given we refer to the 'who' as a DN, and if other values are missing (like the admin details) we will raise a value error on that earlier

File: authentication/active_directory_authentication_provider.py (hunk 2 above)
Comment (adamlavie) [inline authentication/active_directory_authentication_provider.py]: This method is actually used by both the ldap and the AD auth classes and the default values is used as a fallback in-case no specific user_id_attr was specified in the configuration. To clarify, LDAP's default attribute is uid while UPN is the default used by active directory

File: authentication/ldap_authentication_provider.py (hunk 3 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: 1. Actually, all auth methods use bind to authenticate. I wanted to pick a name with more meaning. Maybe auth_by_base_dn would be more clear?
2. The default attr is used in-case no attribute was specified in the initial configuration in self.user_id_attribute

File: authentication/ldap_authentication_provider.py (hunk 4 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: I get what you're saying. totally.

File: authentication/ldap_authentication_provider.py (hunk 5 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: There are actually a lot of result types that can be returned. I'll need to run some tests but it might be a good idea to raise an error if we get a result that's not a search result.

File: authentication/ldap_authentication_provider.py (hunk 6 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: 1. It is indeed surprising but by default, UID is not set to be unique on LDAP (http://stackoverflow.com/questions/13747795/creating-user-accounts-with-unique-uid-in-openldap). There is however an option to configure LDAP so that it would be unique. Regarding the fact that we can raise an error way faster, totally true.
2. Couldn't find anything. What's wrong with search though?

File: authentication/ldap_authentication_provider.py (hunk 8 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: Indeed. It's just how the result is returned from the search. Will be a bit better once I remove the result_list like you recommended above..

File: authentication/ldap_authentication_provider.py (hunk 10 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: the conn.unbind() is being done in the handle_authentication_error function

File: authentication/ldap_authentication_provider.py (hunk 12 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: I though so too but turned out that in this case there is no need for try and catch since if the conn object was created, even if not binded yet but only initialized, the unbind will succeed. I tested it before to verify this. You'll notice that If initialization fails, we don't call on conn.unbind()

File: authentication/ldap_authentication_provider.py (hunk 1 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: very true

File: authentication/ldap_authentication_provider.py (hunk 13 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: This should probably be renamed to create_bind_conn

File: authentication/ldap_authentication_provider.py (hunk 14 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: 1. There is a bigger difference here. It's true that both use bind to authenticate but in the subtree search, the configuration has to hold an admin user and admin password whereas the second method does not require these credentials. What you are suggesting is not possible since searching the tree is only possible if the admin username and password were indeed provided in the configuration.. These things are validated in self.is_subtree_authentication_method()  this would be more clear if you could see the documentation first..
2. If whoever configured the ldap in the blueprint specified a base_dn, it means no one outside this subtree is allowed to login. If a base_dn was not specified then we indeed allow users from the ldap root to login. If the guy configuring the ldap security in the blueprint wants to allow identification using attributes, he should specify the root as a base DN before bootstrapping..

File: authentication/active_directory_authentication_provider.py [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+
+from ldap_authentication_provider import LDAPAuthenticationProvider
+
+DEFAULT_AUTH_ATTR_NAME = 'sAMAccountName'
+
+
+class ActiveDirectoryAuthenticationProvider(LDAPAuthenticationProvider):
+
+    def __init__(self, args, domain_name=None, **kwargs):
+        if 'search_props' in kwargs:
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: sorry for the hassle, but this is user-facing, so we better rename to 'search_properties'

File: authentication/active_directory_authentication_provider.py [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+            if 'user_id_attribute' not in kwargs.get('search_props'):
+                kwargs['search_props']['user_id_attribute'] = \
+                    DEFAULT_AUTH_ATTR_NAME
+        super(ActiveDirectoryAuthenticationProvider, self).__init__(args,
+                                                                    **kwargs)
+        self.domain_name = domain_name
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py:30]: isn't domain name optionally a part of kwargs? does defining it separately in the signature prevents it from being passed to the super's kwargs? if so, good!

File: authentication/active_directory_authentication_provider.py [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+
+    def authenticate(self, userstore=None):
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: is the userstore used here? if not - please remove it from the signature (I know the super has it but it's ok)

File: authentication/ldap_authentication_provider.py [hunk 25]
```diff
@@ -12,35 +12,134 @@
...
     def authenticate(self, userstore=None):
-        username, password = utils.get_basic_http_authentication_info()
+        who, password = self.get_creds_from_request()
+        if self.is_search_based_authentication():
+            self.authenticate_by_attribute(who, password)
+        else:
+            self.validate_bind(who, password)
+        return who
+
+    def authenticate_by_attribute(self, who, password):
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: this method is a utility method. It can be static and should probably be just "get_dn_by_user_attribute(identifying_attribute, value)
and return the user DN.

File: authentication/ldap_authentication_provider.py [hunk 26]
```diff
@@ -12,35 +12,134 @@
...
+            conn = self.create_bind_conn(self.admin_user_id,
+                                         self.admin_password)
+        except AuthenticationException as e:
+            err = 'Failed binding to \'{0}\' using the dedicated ' \
+                  'admin_dn: {1}, provided in the configuration; {2}' \
+                .format(self.ldap_url, who, e)
+            self.lgr.warning(err)
+            raise RuntimeError(err)
+
+        attribute_name = self.user_id_attribute
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: Since this is a utility method it does not have to be aware of properties set on self. also you can reduce the 2-lines to one: search_filter = '{0}={1}'.format(identifying_attribute, value)

File: authentication/ldap_authentication_provider.py [hunk 27]
```diff
@@ -12,35 +12,134 @@
...
+                                       .format(result_type=result_type))
+        except ldap.LDAPError as e:
+            err = 'Failed retrieving results according to result_id; {0}' \
+                .format(e)
+            self.handle_authentication_error(err, conn)
+
+        # unbind admin connection.
+        conn.unbind()
+
+        if len(result_data) != 1:
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: I think that if result_data is none you'll get an exception here

File: authentication/ldap_authentication_provider.py [hunk 28]
```diff
@@ -12,35 +12,134 @@
...
+                    raise RuntimeError('An unexpected result object was '
+                                       'returned by the search: {result_type}'
+                                       .format(result_type=result_type))
+        except ldap.LDAPError as e:
+            err = 'Failed retrieving results according to result_id; {0}' \
+                .format(e)
+            self.handle_authentication_error(err, conn)
+
+        # unbind admin connection.
+        conn.unbind()
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py:130]: unbind should be in finally and not in handle_authenticatio_error, it's un-related to error handling

File: authentication/ldap_authentication_provider.py [hunk 29]
```diff
@@ -12,35 +12,134 @@
...
+
+        try:
+            # Retrieve query results according to the result ID in order to
+            # obtain the user DN to bind with.
+            result_type, result_data = conn.result(ldap_result_id, 1)
+            if result_data:
+                if result_type != ldap.RES_SEARCH_RESULT:
+                    raise RuntimeError('An unexpected result object was '
+                                       'returned by the search: {result_type}'
+                                       .format(result_type=result_type))
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: else:
  # result_data is none
  raise something

File: authentication/ldap_authentication_provider.py [hunk 30]
```diff
@@ -12,35 +12,134 @@
...
+        conn.unbind()
+
+        if len(result_data) != 1:
+            err = 'Expecting a single result matching the user with ' \
+                  'username {0}. Found {1}.' \
+                .format(who, len(result_data))
+            self.handle_authentication_error(err, conn)
+
+        # extract the user DN from the search result.
+        user_dn = result_data[0][0]
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: this is the value to be returned

File: authentication/ldap_authentication_provider.py [hunk 31]
```diff
@@ -12,35 +12,134 @@
...
+
+        # attempt to bind using the matching user ID/DN and provided password.
+        self.validate_bind(user_dn, password)
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: does not belong in this method, should be done even if no search was required

File: authentication/active_directory_authentication_provider.py [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+        who, password = self.get_creds_from_request()
+
+        if self.is_search_based_authentication():
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: we authenticate by binding either way, it's just that we sometimes have to get the user object to retrieve the dn. so, we should ask:
    if self.is_search_based_authentication():
        who = search_and_get_dn(...)

and then just call validate_bind, regardless of search being done before it or not:
self.validate_bind(user_dn, password)

File: authentication/active_directory_authentication_provider.py [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+                           '{user_id} and domain name: {domain}'
+                           .format(user_id=who, domain=self.domain_name))
+            self.authenticate_by_user_id(who, password)
+        return who
+
+    # By default, active directory allows for authentication using the UPN
+    # attribute. If search_props not specified, the implementation will assume
+    # that UPN/other default attr is used and append the domain_name to the
+    # user_id in-case domain_name was specified in the configuration.
+    def authenticate_by_user_id(self, who, password):
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: we should perform validate_bind outside of this method, since we always have to do it regardless of domain or no domain.
This method should can be removed, and you can just use this instead of calling it:
if self.domain_name:
  who = '{0}@{1}'.format(who, self.domain_name)

File: authentication/ldap_authentication_provider.py (hunk 10 above)
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: I understand but don't think it's so great, see above :)

File: authentication/ldap_authentication_provider.py [hunk 34]
```diff
@@ -12,35 +12,134 @@
...
+                .format(who)
+            self.handle_authentication_error(err, conn)
+        except ldap.LDAPError as e:
+            err = 'Failed to bind user \'{0}\'; {1}' \
+                .format(who, e)
+            self.handle_authentication_error(err, conn)
+
+        return conn
+
+    def is_search_based_authentication(self):
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: 1. we need identifying_attribute as well here
2. this is the place to check if some of the 4 mandatory settings are missing or None. if so, we should raise a value error with a detailed message

File: authentication/active_directory_authentication_provider.py [hunk 35]
```diff
@@ -0,0 +1,46 @@
...
+
+
+class ActiveDirectoryAuthenticationProvider(LDAPAuthenticationProvider):
+
+    def __init__(self, ldap_url, domain_name=None, search_properties=None):
+        super(ActiveDirectoryAuthenticationProvider, self)\
+            .__init__(ldap_url, search_properties)
+        self.domain_name = domain_name
+
+    def authenticate(self, userstore=None):
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: if userstore is not used - remove it from the signature. we don't want anyone passing this argument thinking it means something when it's not.

File: authentication/active_directory_authentication_provider.py [hunk 36]
```diff
@@ -0,0 +1,46 @@
...
+        who, password = self.get_creds_from_request()
+
+        if self.search_props:
+            self.search_props.setdefault('user_id_attribute',
+                                         DEFAULT_AUTH_ATTR_NAME)
+            user_dn = self.get_dn_by_user_attribute(who,
+                                                    self.ldap_url,
+                                                    **self.search_props)
+        else:
+            self._logger.debug('Attempting to connect using the user id '
```
Comment (noakup) [inline authentication/active_directory_authentication_provider.py]: 1. if self.domain_name is None you might get an exception here..?
2. I don't see why we only log when search is not used... I think it's better to move the logging to line 44, just before calling validate_bind, and change the message to:
self._logger.debug('Attempting to connect using: {user_dn}'.format(user_dn))
because it includes the domain anyways. also, I removed the 'the user id' part from the message because it might be a UPN or something else.

Comment (adamlavie) [inline authentication/active_directory_authentication_provider.py]: 1. Nop, it will print None

File: authentication/ldap_authentication_provider.py [hunk 37]
```diff
@@ -12,35 +12,156 @@
...
+    _logger = logging.getLogger(constants.FLASK_SECUREST_LOGGER_NAME)
+
+    def __init__(self, ldap_url, search_properties=None):
+        self.ldap_url = ldap_url
+        self.search_props = search_properties
+        self.validate_search_props()
+
+    def validate_search_props(self):
+        if self.search_props:
+            valid = all(test in self.search_props for test in
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: 1. `test` should be renamed to `value` or `element`
2. the condition is wrong. an empty value for a field in the search properties will cause `test in self.search_props` to return `True`, when in fact it is not valid. It should be:
valid = all(search_props.get(value, None) for value in REQ_SEARCH_FIELDS)
3. the valid variable is redundant. you can just put the condition into line 43 like this:
if not all(search_props.get(value, None) for value in REQ_SEARCH_FIELDS):

File: authentication/ldap_authentication_provider.py [hunk 38]
```diff
@@ -12,35 +12,156 @@
...
+                        REQ_SEARCH_FIELDS)
+            if not valid:
+                err = 'The configured search_properties are missing required '\
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: 1. This message has the same problem. An empty value will still cause the message to be something like:
"Search properties must hold values for the following fields: 'a, b, c', got 'a, b, c'

2. brackets of different types will appear in the message:
Search properties must hold values for the following fields: '('base_dn', 'admin_user_id', 'admin_password')', got '['admin_password', 'admin_user_id', 'base_dn']'

File: authentication/ldap_authentication_provider.py (hunk 37 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: 1. There is no harm here
2. This can be a nice improvement
3. You can style your code however you like. I think this is more readable

File: authentication/ldap_authentication_provider.py (hunk 38 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: 1. It's not a problem it's consistent with the implementation. 
2. true, there is no reason to use a list, can be a tuple

File: authentication/ldap_authentication_provider.py [hunk 39]
```diff
@@ -12,35 +12,156 @@
...
+                      'search fields. Search properties must hold values for '\
+                      'the following fields: \'{req_fields}\', got '\
+                      '\'{props_fields}\''\
+                    .format(req_fields=REQ_SEARCH_FIELDS,
+                            props_fields=self.search_props.keys())
+                self._logger.warning(err)
+                raise RuntimeError(err)
+
+    # LDAP authentication can be done by subtree match or according to
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: the comment is wrong, it's always done by binding..

File: authentication/ldap_authentication_provider.py [hunk 40]
```diff
@@ -12,35 +12,156 @@
...
+        self.validate_bind(user_dn, password)
+        return who
+
+    @classmethod
+    def handle_authentication_error(cls, err, conn):
+        conn.unbind()
+        cls._logger.warning(err)
+        raise AuthenticationException(err)
+
+    def validate_bind(self, who, password):
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: utility method, you can pass self.ldap_url as the first arg and then make it a classmethod

File: authentication/ldap_authentication_provider.py (hunk 39 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: ??

File: authentication/ldap_authentication_provider.py [hunk 41]
```diff
@@ -12,35 +12,156 @@
...
+                                        admin_user_id,
+                                        admin_password)
+            # Search for matches under the base_dn subtree using the attribute
+            # search filter.
+            ldap_result_id = conn.search(base_dn,
+                                         ldap.SCOPE_SUBTREE,
+                                         search_filter,
+                                         [DN_ATTRIBUTE_NAME])
+        except AuthenticationException as e:
+            err = 'Failed binding to \'{0}\' using the dedicated ' \
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: seems like you should call `handle_authentication_error` and pass None as the `conn` argument.
In `handle_authentication_error` you should call `conn.unbind()` only if conn is not None.
That way any error handling actions added into `handle_authentication_error` will not be missing / duplicated here

File: authentication/ldap_authentication_provider.py (hunk 40 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: I can't do that since this method is being called from class methods that don't hold a reference to the actual instance

Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: I found it to be more clear this way..

File: authentication/ldap_authentication_provider.py [hunk 42]
```diff
@@ -12,35 +12,156 @@
...
+            # Retrieve query results according to the result ID in order to
+            # obtain the user DN to bind with.
+            result_type, result_data = conn.result(ldap_result_id, 1)
+            if result_data:
+                if result_type != ldap.RES_SEARCH_RESULT:
+                    raise RuntimeError('An unexpected result object was '
+                                       'returned by the search: {result_type}'
+                                       .format(result_type=result_type))
+            else:
+                raise RuntimeError('An empty result was returned by the search'
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: an empty result is a bit of a developer-oriented message. We can rephrase it to something like:
`User \'{user_id}\' not found.format...`

File: authentication/ldap_authentication_provider.py (hunk 41 above)
Comment (adamlavie) [inline authentication/ldap_authentication_provider.py]: No this throws a RuntimeError not auth error

File: authentication/ldap_authentication_provider.py [hunk 43]
```diff
@@ -12,35 +12,156 @@
...
+        except ldap.LDAPError as e:
+            err = 'Failed retrieving results according to result_id; {0}' \
+                .format(e)
+            cls.handle_authentication_error(err, conn)
+
+        # unbind admin connection.
+        conn.unbind()
+
+        if len(result_data) != 1:
+            err = 'Expecting a single result matching the user with ' \
```
Comment (noakup) [inline authentication/ldap_authentication_provider.py]: "matching the user with username" ->> "matching username"

File: authentication/active_directory_authentication_provider.py (hunk 35 above)
Comment (adamlavie) [inline authentication/active_directory_authentication_provider.py]: Sadly, this can't be done because of the way the authentication method is being invoked
